### PR TITLE
Raise an error when empty list is passed as wires for operators with `AnyWires` and `AllWires`

### PR DESCRIPTION
**Pull request by meenu-kri**
**Context:** Operators that have `num_wires=AnyWires` or `AllWires` were applied to all wires when instantiated with wires = []. This is not desirable for most operators except a few, such as, `qml.Barrier` and `qml.Snapshot`.

**Description of the Change:** Raise an error for operators that have `num_wires=AnyWires` or `AllWires` when instantiated with an empty list for `wires` except the following operators: `qml.Barrier`, `qml.Snapshot`, and `qml.Hamiltonian`.

**Benefits:**

**Possible Drawbacks:** 

**Related GitHub Issues:** https://github.com/PennyLaneAI/pennylane/issues/2829

### CODE REVIEW

File: pennylane/operation.py [hunk 1]
```diff
@@ -891,6 +891,15 @@ def __init__(self, *params, wires=None, do_queue=True, id=None):
...
             )
 
+        # Check that the wires is not an empty list for operators that have
+        # `num_wires` as `AnyWires` or `AllWires`.
+        if self.num_wires in {AllWires, AnyWires}:
+            if not isinstance(self, (qml.Barrier, qml.Snapshot, qml.Hamiltonian)):
+                # Barrier, Snapshot: Applied to all wires if instantiated with wires = [].
+                # Hamiltonian: Possible to be empty with simplify().
+                if len(qml.wires.Wires(wires)) == 0:
+                    raise ValueError("require valid wires.")
```
Comment (AlbertMitjans) [inline pennylane/operation.py:901]: The error message is too cryptic. Can we give a bit more information?

**Review by AlbertMitjans — COMMENTED** [review]
Good job!! 🎉 🚀 

Just one small comment.

File: pennylane/operation.py (hunk 1 above)
Comment (meenu-kri) [reply pennylane/operation.py]: Sure. I made the following change to the error message.
```if len(qml.wires.Wires(wires)) == 0:
                    raise ValueError(
                        f"{self.name}: wrong number of wires. "
                        f"At least one wire has to be given."
                    )
```

File: pennylane/operation.py [hunk 2]
```diff
@@ -891,6 +891,18 @@ def __init__(self, *params, wires=None, do_queue=True, id=None):
                 f"{len(self._wires)} wires given, {self.num_wires} expected."
             )
 
+        # Check that the wires is not an empty list for operators that have
+        # `num_wires` as `AnyWires` or `AllWires`.
+        if self.num_wires in {AllWires, AnyWires}:
+            if not isinstance(self, (qml.Barrier, qml.Snapshot, qml.Hamiltonian)):
```
Comment (antalszava) [inline pennylane/operation.py:897]: Is there are test cases already for each of these operations testing that no error is raised when passing e.g., an empty list as wires?

**Review by antalszava — COMMENTED** [review]
Looking good overall, one question! :tada:

File: pennylane/operation.py (hunk 2 above)
Comment (meenu-kri) [reply pennylane/operation.py:897]: That's a good point! I have now added test cases for `qml.Barrier`, `qml.Snapshot`, `qml.Hamiltonian` testing that no 'empty wire list' error is raised for these cases.

**Review by antalszava — APPROVED** [review]
Looks good! :tada:

**Comment by antalszava** [comment]
@AlbertMitjans once this looks good to you, could you merge it?

**Pull request merged**